src/sctools/bam.py

```diff
@@ -226,19 +234,150 @@ def tag(self, output_bam_name: str, tag_generators) -> None:
+                barcodes.add(barcode)
+    return barcodes
+
+
+def get_barcode_for_alignment(
+    alignment: pysam.AlignedSegment, tags: List[str], raise_missing: bool
+) -> str:
+    """ Get the barcode for an Alignment
+
+    :param alignment: pysam.AlignedSegment
+        An Alignment from pysam.
+    :param tags: List[str]
+        Tags in the bam that might contain barcodes.
+    :param raise_missing: bool
+        Raise an error if no barcodes can be found.
+    :return: str
+        A barcode for the alignment, or None if one is not found and raise_missing is False.
+    """
+    alignment_barcode = None
+    for tag in tags:
+        # The non-existent barcode should be the exceptional case, so try/except is faster than if/else
+        try:
+            alignment_barcode = alignment.get_tag(tag)
+            break  # Got the key, don't bother getting the next tag
+        except KeyError:
+            continue  # Try to get the next tag
+
+    if raise_missing and alignment_barcode is None:
+        raise RuntimeError(
+            'Alignment encountered that is missing {} tag(s).'.format(tags)
+        )
+
+    return alignment_barcode
+
+
+def write_barcodes_to_bins(
+    in_bam: str, tags: List[str], barcodes_to_bins: Dict[str, int], raise_missing: bool
+) -> List[str]:
+    """ Write barcodes to appropriate shards as defined by barcodes_to_bins
+
+    :param in_bam: str
+        The bam file to read.
+    :param tags: List[str]
+        Tags in the bam that might contain barcodes.
+    :param barcodes_to_bins: Dict[str, int]
+        A Dict from barcode to bin. All barcodes of the same type need to be written to the same shard.
+        These numbered shards are merged after parallelization so that all alignments with the same
+        barcode are in the same bam.
+    :param raise_missing: bool
+        Raise an error if no barcodes can be found.
+    :return: A list of paths to the written shards.
+    """
+    # Create all the output files
+    with pysam.AlignmentFile(in_bam, 'rb', check_sq=False) as input_alignments:
+
+        # We need a random int appended to the dirname to make sure input bams with the same name don't clash
+        dirname = (
+            os.path.splitext(os.path.basename(in_bam))[0]
+            + "_"
+            + str(random.randint(0, 10000))
+        )
+        os.makedirs(dirname)
+
+        files = []
+        bins = list(set(barcodes_to_bins.values()))
+        filepaths = []
+        for i in range(len(bins)):
+            out_bam_name = os.path.realpath(dirname) + ("/" + dirname + '_%d.bam' % i)
+            filepaths.append(out_bam_name)
+            # For now, bam writing uses one thread for compression. Better logic could support more threads without
+            # starving the machine for resources
+            open_bam = pysam.AlignmentFile(
+                out_bam_name, 'wb', template=input_alignments
+            )
+            files.append(open_bam)
```

Comment:
I don't think it's that fragile. Creating a `dict` seems like overkill because its only doing the work of an indexed array. Would a comment explaining the relationship between `barcodes_to_bins` and `files` suffice?

Reply:
Sure, that works for me.